server/src/main/java/com/thoughtworks/go/config/PartialConfigService.java

```diff
@@ -69,27 +70,57 @@ public void onSuccessPartialConfig(ConfigRepoConfig repoConfig, PartialConfig in
 
-                if (updateConfig(incoming, fingerprint, repoConfig)) {
+                //validate rules
+                hasRuleViolations(incoming);
+
+                //validate config.
+                // updateConfig will fail to update the configuration if there are validation errors.
+                // Even in case of rules violation, the updateConfig method is required to populate a server health message of rule violation, which also will be shown on the config repo spa.
+                final boolean isConfigUpdated = updateConfig(incoming, fingerprint, repoConfig);
+
+                if (isConfigUpdated) {
+                    // mark the partial as valid when config is updated successfully for it.
                     cachedGoPartials.markAsValid(fingerprint, incoming);
                 } else {
                     final PartialConfig previousValidPartial = cachedGoPartials.getValid(repoConfig.getRepo().getFingerprint());
 
-                    if (violatesRules && hasRuleViolations(previousValidPartial)) {
-                        // do not allow fallback to the last version of the partial if the current rules do not allow
-                        cachedGoPartials.removeValid(repoConfig.getRepo().getFingerprint());
+                    if (previousValidPartial != null) {
+                        final boolean isSameRevisionAsOfPreviousValidPartial = ((RepoConfigOrigin) incoming.getOrigin()).getRevision().equals(((RepoConfigOrigin) previousValidPartial.getOrigin()).getRevision());
+                        // In case of last valid partials are same as of incoming partials, config repo rules are changed causing the rule validation failure.
+                        // in such case, remove the valid partials without clearing the server health message. Server health message is populated for the same fingerprint with the latest parse failure message.
+                        if (isSameRevisionAsOfPreviousValidPartial) {
+                            removeCachedPartialsAndUpdateConfig(fingerprint);
+                        } else {
+                            // lets say the latest partial is invalid for the current config repo rules.
+                            // in such cases, validate the previous valid partial with respect to current config repo rules.
+                            // if the previous valid partials are valid - do nothing - as the error for the latest partial is already populated and config contains the last known partial
+                            // and if the previous valid partials are invalid - remove those config without clearing the server health message. Server health message is populated for the same fingerprint with the latest parse failure message.
+                            ((RepoConfigOrigin) previousValidPartial.getOrigin()).setConfigRepo(((RepoConfigOrigin) incoming.getOrigin()).getConfigRepo());
+                            if (hasRuleViolations(previousValidPartial)) {
```

Comment:
I get your point why we need to check rules for last valid partial against the current rules. guess my previous comment has to change to something like,

```
if (updateConfig(incoming, fingerprint, repoConfig)) {
    cachedGoPartials.markAsValid(fingerprint, incoming);
} else {
    final PartialConfig previousValidPartial = cachedGoPartials.getValid(repoConfig.getRepo().getFingerprint());
    if (validPartialHasRuleViolations(previousValidPartial)) {
        // do not allow fallback to the last version of the partial if the current rules do not allow
        cachedGoPartials.removeValidWithoutClearingServerHealthMessage(repoConfig.getRepo().getFingerprint());
        goConfigService.updateConfig(cruiseConfig -> {
            cruiseConfig.getPartials().remove(cachedGoPartials.findPartialByFingerprint(cruiseConfig, fingerprint));
            return cruiseConfig;
        });
    }
}
```